test/dummies/BadRelayClient.ts

```diff
@@ -0,0 +1,37 @@
+import RelayClient, { RelayingResult, TmpDependencyTree } from '../../src/relayclient/RelayClient'
+import GsnTransactionDetails from '../../src/relayclient/types/GsnTransactionDetails'
+import { Address } from '../../src/relayclient/types/Aliases'
+import { RelayClientConfig } from '../../src/relayclient/GSNConfigurator'
+
+export default class BadRelayClient extends RelayClient {
```

Comment:
its only for tests, so can be achieved much simpler. e.g.
```
badCliet = new RelayClient()
badCleint.relayTransaction = ()=>{throw new Error()}}
```

Reply:
I have a very strong opinion against assigning methods of objects. You are not allowed (at least without reflection) to do it in any sane language. 
The fact you can do it in JS/TS does not mean you should.

Reply:
I agree with Dror, this is one of the few good things about testing in js.

Reply:
It is good until it is bad. Your `badClient` no longer implements it's own class or interface, it is just object that has `relayTransaction = ()=>{}` and that's it. If you remove the entire `relayTransaction` from the original object, or rename it to 'relayTra**sn**action' again, test will still pass. 
With TypeScript if you define `badClient` as `RelayClient` it will fail in compile-time with method not found.

Reply:
the only place I think its good is when writing tests. tests should be self-contained: having a utilty class merely to "stub" a method to return an error makes the test hard to read.  Your construction of `BadRelayClient()` gets some boolean params, which are meaningless when reading the test. its better to assign a method - or use `sinon`

Reply:
Future: we need guidelines on how to write tests that need to modify behaviour of existing objects (subclass with modified methods, stubbing/mocking existing ones, plain modification inside a test, etc)